File: node/src/aktcache.ts
import AsyncLock from "async-lock";
import { Cache, newCache } from "./cache";
import { Entry } from "./lru/list";
import { PeerGetter, PeerPicker } from "./peers";
import { SingleFlight } from "./singleflight/singleflight";

const AKTCACHENAMEMAP: { [propsName: string]: Group } = {};

const GLOBALLOCK = new AsyncLock();

export interface Getter {
  Get(key: string): any | Promise<any> | null;
}

export class Group {
  name: string;
  getter: Getter;
  mainCache: Cache;
  peers: PeerPicker;
  loader: SingleFlight;
  constructor(name: string, getters: Getter, mainCache: Cache) {
    name = `__AKTCACHE__${name}`;
    this.name = name;
    this.getter = getters;
    this.mainCache = mainCache;
    this.loader = new SingleFlight(name);
    AKTCACHENAMEMAP[name] = this;
  }

  registerPeers(peers: PeerPicker) {
    if (!!this.peers) {
      throw new Error("egisterPeerPicker called more than once");
    }
    this.peers = peers;
  }

  async get(key: string): Promise<any | null> {
    if (!key) {
      return null;
    }
    try {
      const v = await this.mainCache.get(key);
      if (v !== null) {
        return v.value;
      }
      const loadv = await this.load(key);
      return loadv ? loadv.value : null;
    } catch (e) {
      console.log(e.toString());
      return null;
    }
  }

  async load(key: string): Promise<Entry | null> {
    try {
      return await this.loader.Do(key, async () => {
        if (!this.peers) {
          const peer = await this.peers.pickPeer(key);
          if (peer) {
            const value: Entry | null = await this.getFromPeer(peer, key);
            if (value) {
              return value;
            }
          }
        }
        return await this.getLocall(key);
      });
    } catch (e) {
      console.log(e.toString());
      return null;
    }
  }

  async getFromPeer(peer: PeerGetter, key: string) {
    const value = await peer.get(this.name, key);
    if (!value) {
      return null;
    }
    return new Entry(key, value);
  }

  async getLocall(key: string): Promise<Entry | null> {
    try {
      const value = await this.getter.Get(key);

      if (!value) {
        return null;
      }

      await this.populateCache(key, value);

      return new Entry(key, value);
    } catch (e) {
      console.log(e.toString());
      return null;
    }
  }

  private async populateCache(key: string, v: any) {
    try {
      await this.mainCache.add(key, v);
    } catch (e) {
      console.log(e.toString());
      return null;
    }
  }
}

export const newGroup = async (
  name: string,
  getters: Getter,
  cacheBytes: number
) => {
  try {
    const group = await GLOBALLOCK.acquire(
      "___AKTCGROUP___NEW___",
      async () => {
        const mapName = `__AKTCACHE__${name}`;
        if (AKTCACHENAMEMAP[mapName]) {
          throw Error("this name is exist");
        }
        const cache = await newCache(`__GROUP__CREATE__${name}`, cacheBytes);
        if (!cache) {
          throw Error("create cache failed");
        }
        return new Group(name, getters, cache);
      }
    );
    return group;
  } catch (e) {
    console.log(e.toString());
    return null;
  }
};

export const getGroup = async (name: string) => {
  try {
    const group = GLOBALLOCK.acquire("___AKTCGROUP___GET___", () => {
      const mapName = `__AKTCACHE__${name}`;
      return AKTCACHENAMEMAP[mapName] || null;
    });
    return group;
  } catch (e) {
    console.log(e.toString());
    return null;
  }
};

File: go/aktcache/aktcache.go
package aktcache

import (
	"fmt"
	"sync"
)

type Getter interface {
	Get(key string) ([]byte, error)
}

type GetterFunc func(key string) ([]byte, error)

func (f GetterFunc) Get(key string) ([]byte, error) {
	return f(key)
}

type Group struct {
	name      string // 名称
	getter    Getter // 缓存未命中时的回调
	mainCache cache  // 并发缓存
	peers     PeerPicker
}

var (
	mu     sync.RWMutex
	groups = make(map[string]*Group)
)

func NewGroup(name string, cacheBytes int64, getter Getter) *Group {
	if getter == nil {
		panic("nil Getter")
	}

	mu.Lock()
	defer mu.Unlock()

	g := &Group{
		name:   name,
		getter: getter,
		mainCache: cache{
			cacheBytes: cacheBytes,
		},
	}

	groups[name] = g
	return g
}

func GetGroup(name string) *Group {
	mu.RLock()
	defer mu.RUnlock()
	g := groups[name]
	return g
}

func (g *Group) RegisterPeers(peers PeerPicker) {
	if g.peers != nil {
		panic("RegisterPeerPicker called more than once")
	}
	g.peers = peers
}

// 查询value
func (g *Group) Get(key string) (ByteView, error) {
	if key == "" {
		return ByteView{}, fmt.Errorf("key is required")
	}

	// 如果mainCache存在缓存，则直接返回
	if v, ok := g.mainCache.get(key); ok {
		return v, nil
	}

	// 如果不存在
	return g.load(key)
}

//
func (g *Group) load(key string) (value ByteView, err error) {

	if g.peers != nil {
		if peer, ok := g.peers.PickPeer(key); ok {
			if value, err := g.getFromPeer(peer, key); err == nil {
				return value, nil
			}
			fmt.Println("[GeeCache] Failed to get from peer", err)
		}
	}

	// 单机时调用本地Get回调函数
	return g.getLocally(key)
}

func (g *Group) getFromPeer(peer PeerGetter, key string) (ByteView, error) {
	bytes, err := peer.Get(g.name, key)
	if err != nil {
		return ByteView{}, err
	}
	return ByteView{b: bytes}, nil
}

// 单机不存在时，产生原数据，缓存至mainCache
func (g *Group) getLocally(key string) (ByteView, error) {
	bytes, err := g.getter.Get(key)
	if err != nil {
		return ByteView{}, err
	}

	value := ByteView{b: cloneBytes(bytes)}
	g.populateCache(key, value)
	return value, nil
}

func (g *Group) populateCache(key string, value ByteView) {
	g.mainCache.add(key, value)
}
